File: leetcode/39.CombinationSum.go
package leetcode

func combinationSum(candidates []int, target int) [][]int {
	var result [][]int

	var traverse func(i int, current []int, total int)
	traverse = func(i int, current []int, total int) {
		if total == target {
			result = append(result, append([]int{}, current...))
			return
		}

		if i >= len(candidates) || total > target {
			return
		}

		current = append(current, candidates[i])
		traverse(i, current, total+candidates[i])

		current = current[:len(current)-1]
		traverse(i+1, current, total)
	}

	traverse(0, []int{}, 0)

	return result
}

File: leetcode/39.CombinationSum.ts
function combinationSum(candidates: number[], target: number): number[][] {
  const result: number[][] = []

  const traverse = (i: number, current: number[], total: number) => {
    if (total === target) {
      result.push([...current])
      return
    }

    if (i >= candidates.length) return
    if (total > target) return

    current.push(candidates[i])
    traverse(i, current, total + candidates[i])

    current.pop()
    traverse(i + 1, current, total)
  }

  traverse(0, [], 0)

  return result
}

console.log(combinationSum([2, 3, 6, 7], 7))
